lib/kcl.js

```diff
@@ -215,19 +218,23 @@ module.exports = function(config, kinesis) {
         throttledPutToCloudwatch('ShardIteratorAgeInMs', (+new Date()) - resp.Records[0].ApproximateArrivalTimestamp, 'Milliseconds', shard.id);
       }
 
-      // malformed responses will have an undefined NextShardIterator. In that case, we keep the existing one.
-      shard.iterator = resp.data.NextShardIterator === undefined ? shard.iterator : resp.data.NextShardIterator;
+      // malformed responses will have an undefined NextShardIterator and undefined Records key
+      // only replace the iterator if we have a nextShardIterator. otherwise we'll keep using it
+      if (resp.data.NextShardIterator) shard.iterator = resp.data.NextShardIterator;
       shard.lastGetRecords = +new Date();
 
-      if (resp.data.Records && resp.data.Records.length > 0) {
+      if (resp.data.Records && resp.data.Records.length > 0 && shard.iterator) {
         shard.sequenceNumber = resp.data.Records[resp.data.Records.length - 1].SequenceNumber;
         config.processRecords.call(shard, resp.data.Records, processRecordsDone);
       } else {
-        if (shard.iterator === null) {
+        // if the shard iterator is undefined (or null...) and the response is valid (not a malformed payload, with undefined Records key)
+        if (!resp.data.NextShardIterator && resp.data.Records) {
           db.shardComplete(shard, function (err) {
             if (err) throw err;
+            console.log('Shard', shard.id, 'marked as complete');
```

Comment:
So I figured...since we do [log for rare and exceptional events](https://github.com/mapbox/kine/blob/master/lib/kcl.js#L323), perhaps kine could do it here too...